[TNL-2896] Colons in problem names

vkaracic opened the pull request:
WORK IN PROGRESS

- [x] `replace_static_urls` test
- [x] `grade_histogram` test
- [x] create `setUp` method for DRY
- [x] remove unused imports
- [x] change course keys to new ones
- [x] alphabetize imports
- [x] squash commits

https://openedx.atlassian.net/browse/TNL-2896

edx-webhook commented:
Thanks for the pull request, @vkaracic! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-platform&number=10352

adampalay commented on openedx/core/lib/tests/test_xblock_utils.py:
```diff
@@ -0,0 +1,202 @@
+
+        self.content_store = contentstore()
+        # pylint: disable=protected-access
+        self.module_store = modulestore()._get_modulestore_by_type(ModuleStoreEnum.Type.mongo)
```
Is there a reason you're specifying this to the mongo modulestore? Shouldn't this really work in both mongo and split? (And if you're testing one, you should test split)

vkaracic commented:
test this please

vkaracic commented:
jenkins test all

vkaracic commented:
jenkins run all

vkaracic commented:
@adampalay I keep getting this error:
```
   Traceback (most recent call last):
    openedx/core/lib/tests/test_xblock_utils.py line 186 in test_add_staff_markup
      problem = self.module_store.get_item(course_key.make_usage_key('problem', 'haveacolon'))
    common/lib/xmodule/xmodule/modulestore/split_mongo/split_draft.py line 266 in get_item
      return super(DraftVersioningModuleStore, self).get_item(usage_key, depth=depth, **kwargs)
    common/lib/xmodule/xmodule/modulestore/split_mongo/split.py line 1055 in get_item
      items = self._load_items(course, [BlockKey.from_usage_key(usage_key)], depth, **kwargs)
    common/lib/xmodule/xmodule/modulestore/split_mongo/split.py line 764 in _load_items
      return [runtime.load_item(block_key, course_entry, **kwargs) for block_key in block_keys]
    common/lib/xmodule/xmodule/modulestore/split_mongo/caching_descriptor_system.py line 127 in _load_item
      block_data = self.get_module_data(block_key, course_key)
    common/lib/xmodule/xmodule/modulestore/split_mongo/caching_descriptor_system.py line 148 in get_module_data
      raise ItemNotFoundError(block_key)
   ItemNotFoundError: BlockKey(type='problem', id='haveacolon')
```
What am I doing wrong?

adampalay commented:
hey @vkaracic , can I have some context please? When are you running into that error?

vkaracic commented:
@adampalay sure. I'm getting that error here:
https://github.com/edx/edx-platform/pull/10352/files#diff-39e5ff091dc2fd6fd5f523d06fe8268eR183

I'm importing a course that has a problem which contains a colon in it's name. You suggested using ``get_item`` to get the element id of a problem, but I'm getting that error. Obviously I'm doing something wrong, but I'm having a hard time finding the right usage of ``get_item`` anywhere so I'm stuck.

adampalay commented:
hey @vkaracic , sorry I didn't have a chance to look at this today; I got caught up with other stuff. I'll be able to take a better look tomorrow.

Adam

adampalay commented on openedx/core/lib/tests/test_xblock_utils.py:
```diff
@@ -0,0 +1,195 @@
+TEST_DATA_DIR = settings.COMMON_TEST_DATA_ROOT
+
+
+class TestXblockUtils(ModuleStoreTestCase):
```
It would be great to test this against both modulestores. We have a pattern to do this using `ddt`. See, for example:

https://github.com/edx/edx-platform/blob/master/cms/djangoapps/contentstore/management/commands/tests/test_create_course.py#L47-L70

adampalay commented on openedx/core/lib/tests/test_xblock_utils.py:
```diff
@@ -0,0 +1,195 @@
+
+        self.content_store = contentstore()
+        # pylint: disable=protected-access
+        self.module_store = modulestore()._get_modulestore_by_type(ModuleStoreEnum.Type.split)
```
I don't know why you're specifying this — this test should be modulestore-agnostic (in fact, you should run it in both `split` and `mongo` modulestores

adampalay commented on openedx/core/lib/tests/test_xblock_utils.py:
```diff
@@ -0,0 +1,195 @@
+            create_if_not_present=True
+        )
+        course_key = course[0].id
+        problem = self.module_store.get_item(course_key.make_usage_key('problem', 'haveacolon'))
```
your problem here is that the usage key should be made with the colon in its name:
```python
usage_key = course_key.make_usage_key('problem', 'I:haveacolon')
problem = modulestore().get_item(usage_key)
```

vkaracic replied:
Thanks! I didn't go as far as to test it with both modulestores, my first priority here is to successfully test the colon problem fix in one of the modulestores then I can go adjust whatever is needed.
Also, I couldn't find tests for any of the functions here so I went ahead and created for those as well, I hope I'm not duplicating tests here.

adampalay replied:
Oh no, we love tests :). Out of curiousity, does this PR boost our coverage? That could be a good indicator of whether or not those functions were covered.

And re testing in modulestore, that makes sense. I'll hang back then until you're ready for a code review. But please do let me know if I can otherwise be helpful.

vkaracic replied:
Now that you have mentioned it, I'm having trouble with the paver coverage commands. How can I see if this PR boosts coverage?

adampalay replied:
@vkaracic you may be able to check coveralls: https://coveralls.io/github/edx/edx-platform

vkaracic replied:
@adampalay if I read this correctly it rose .31%
https://coveralls.io/github/edx/edx-platform?branch=vkaracic%2FTNL-2896

vkaracic commented:
jenkins run lettuce

vkaracic commented:
@adampalay I changed the logic here a little bit. I've moved the two consecutive ``replace`` to a new function that does the replacements with regex, and should there be a need to replace anything else it's just a matter of adding a regex pattern. This enabled me to do a far simpler unit test on the fix for this, which was the reason for this PR in the first place. 

There are only two functions here left to test( ``add_staff_markup`` and ``get_course_update_items`` ) and unfortunately I don't have enough knowledge about their workings to create a proper test for them, nor do I have the time for it right now. So maybe create a ticket for someone from TNL to finish this.

Please review this code, and let me know if I should change anything. Also can you please recommend someone to tag for a second review? I would really like to close this (LOOONG lasting) ticket already :)

adampalay commented on openedx/core/lib/tests/test_xblock_utils.py:
```diff
@@ -0,0 +1,155 @@
+        test_replace = replace_static_urls(
+            data_dir=None,
+            course_id=COURSE_KEY,
+            block=CourseFactory.create(),
```
I think you need to pass the `store` param to the `CourseFactory.create` method: I think it should be like

`CourseFactory.create(default_store=store)`

adampalay replied:
The better strategy here might be to make the course upfront in a `setUpClass` class method. Particularly because you're not changing the course at all. For more information on how to do this, check out this blog post: https://open.edx.org/blog/writing-faster-modulestore-tests

This will also allow you not to have to make that `COURSE_KEY` constant, which is an old mongo course key.

adampalay replied:
You'll still have to, in each test, test across modulestores, so I'm guessing you'll want to create one split course and one old mongo course in the setUpClass method, and then ddt across them in each tests.

adampalay commented on openedx/core/lib/tests/test_xblock_utils.py:
```diff
@@ -0,0 +1,155 @@
+)
+
+COURSE_KEY = CourseKey.from_string('TestX/TS01/2015')
+TEST_DATA_DIR = settings.COMMON_TEST_DATA_ROOT
```
what's this for? I don't see you using this anywhere else.

adampalay commented on openedx/core/lib/tests/test_xblock_utils.py:
```diff
@@ -0,0 +1,155 @@
+        dirty_string = 'I:have-un:allowed_characters'
+        clean_string = sanitize_html_id(dirty_string)
+
+        self.assertEqual(clean_string, 'I_have_un_allowed_characters')
```
possible to have a collision, but I guess that was possible before too :(

vkaracic commented:
@adampalay, everything is addressed. Please take a look, also please recommend someone from TNL to do a second review on this.
Thanks.

robrap commented on openedx/core/lib/xblock_utils.py:
```diff
@@ -5,9 +5,10 @@
 import datetime
 import json
 import logging
+import markupsafe
```
doesn't look like markupsafe got used.

robrap commented on openedx/core/lib/xblock_utils.py:
```diff
@@ -211,6 +212,12 @@ def grade_histogram(module_id):
 
 
+def sanitize_html_id(html_id):
+    """Template uses element_id in js function names, so can't allow dashes and colons"""
```
Comment should be on its own line.  I need to do some more research, but I'm not clear on where the html_id is getting defined.  It seems like there are many more illegal characters for js function names, if that is the root of the problem.  Is there a way to avoid the problem altogether?  Otherwise, we'll need a more complicated sanitize method:
- https://mothereff.in/js-variables
- http://stackoverflow.com/questions/1661197/what-characters-are-valid-for-javascript-variable-names
Functionas and variables might have different rules, so this may not be perfect, but you get the point hopefully.

vkaracic replied:
Hi @robrap thanks for taking the time to review this PR.
For when ``html_id`` contained dashes as special characters there was a ``replace`` in place down at line: https://github.com/edx/edx-platform/pull/10352/files#diff-fd85edee8483bf8cc2814b8d453d12d1L303 
MIT folks, as I've been told, tend to create their courses as XML files and import them that way. When you would group, for example problems, into one folder you would get a structure like this:
```
problem group
    | - problem1
    | - problem2
  . . .
```
and that was reflected in the ID's as 
```
problem_group:problem1
problem_group:problem2
. . .
```
The rendering of that colon was a problem because a link would stop there and wouldn't function. That's why the replacement was necessary, also only dashes and colons are likely to be present in these ID's and therefor only the two are replaced, but I've left it open to add additional characters if they are found in the future.
For more context, please checkout the discussion in the comments here: https://openedx.atlassian.net/browse/TNL-2896

Also, I don't understand the "comment should be on its own line" comment.

robrap replied:
Thanks for the explanation @vkaracic.

Here are our Python guidelines: https://github.com/edx/edx-platform/wiki/Python-Guidelines
The one-line comment here, and through the rest of the PR, should look as follows:
```
"""
Template uses element_id in js function names, so can't allow dashes and colons
"""
```

I imagine the html_id doesn't need to be reversible then, and hopefully it remains unique enough to serve its purpose.

robrap commented on openedx/core/lib/xblock_utils.py:
```diff
@@ -145,7 +146,7 @@ def wrap_xblock(
     """
     This will replace a link between courseware in the format
-    /jump_to/<id> with a URL for a page that will correctly redirect
+    /jump_to_id/<id> with a URL for a page that will correctly redirect
```
Line 160 also used "jump_to".

cahrens commented:
I can take a look at this tomorrow.

cahrens commented:
:+1: Thanks for adding the tests!

robrap commented:
I also meant to thank you for the new tests as well and forget.  It is much appreciated.

vkaracic commented:
Glad to help out! :) Hopefully this will boost the coverage to 95%!
I've made the change to the docstring @robrap please take a second look and let me know if you find anything.

Waiting on @adampalay final review and then I'll merge this.

robrap commented:
@vkaracic The test class has the same docstring issue.  Sorry I didn't make that more clear.  Other than that, I'm good with a :+1: .

robrap commented:
jenkins run lettuce

adampalay commented on openedx/core/lib/tests/test_xblock_utils.py:
```diff
@@ -0,0 +1,208 @@
+        self.assertIsInstance(test_wrap_output, Fragment)
+        self.assertIn('xblock-baseview', test_wrap_output.content)
+        self.assertIn('data-runtime-class="TestRuntime"', test_wrap_output.content)
+        if course_id == 'course_mongo':
```
the better pattern for this is to include the `data-usage-id` and then use `ddt.unpack`: http://ddt.readthedocs.org/en/latest/example.html.

So it would look like:

```python
@ddt.data(
    ('course_mongo',  'data-usage-id="i4x:;_;_TestX;_TS01;_course;_2015"'),
    ('course_split', 'data-usage-id="block-v1:TestX+TS02+2015+type@course+block@course"')
)
@ddt.unpack
def test_wrap_xblock(self, course_id, data-usage-id):
    ...
```

I'm happy to add this as a commit myself, if you'd like :)

vkaracic replied:
Good catch! I'm going to fix this.

adampalay commented on openedx/core/lib/tests/test_xblock_utils.py:
```diff
@@ -0,0 +1,208 @@
+            context=None
+        )
+        self.assertIsInstance(test_replace, Fragment)
+        if course_id == 'course_mongo':
```
see comment above for a nicer way to do this

adampalay commented on openedx/core/lib/tests/test_xblock_utils.py:
```diff
@@ -0,0 +1,208 @@
+            context=None
+        )
+        self.assertIsInstance(test_replace, Fragment)
+        if course_id == 'course_mongo':
```
ditto

adampalay commented:
@vkaracic , I have some code style suggestions for the tests, but once those are in, this is looking great! Thank you for your persistence!

adampalay commented:
@vkaracic , thank you! :+1: !

adampalay commented:
congrats! :fireworks:

vkaracic commented:
Thanks! It's been a long and hard battle :D
